cmd/bosun/sched/check.go

```diff
@@ -200,58 +186,135 @@ func (s *Schedule) RunHistory(r *RunHistory) {
 	s.Save()
 }
 
-// CheckUnknown checks for unknown alerts.
-func (s *Schedule) CheckUnknown() {
-	for range time.Tick(s.Conf.CheckFrequency / 4) {
-		log.Println("checkUnknown")
-		r := s.NewRunHistory(time.Now(), nil)
-		s.Lock()
-		for ak, st := range s.status {
-			if st.Forgotten {
-				continue
-			}
-			a := s.Conf.Alerts[ak.Name()]
-			t := a.Unknown
-			if t == 0 {
-				t = s.Conf.CheckFrequency * 2
-			}
-			if t == 0 {
-				continue
-			}
-			if time.Since(st.Touched) < t {
-				continue
-			}
-			r.Events[ak] = &Event{Status: StUnknown}
+func (r *RunHistory) GetUnknownAndUnevaluatedAlertKeys(alert string) (unknown, uneval []expr.AlertKey) {
+	unknown = []expr.AlertKey{}
+	uneval = []expr.AlertKey{}
+	for ak, ev := range r.Events {
+		if ak.Name() != alert {
+			continue
+		}
+		if ev.Status == StUnknown {
+			unknown = append(unknown, ak)
+		} else if ev.Unevaluated {
+			uneval = append(uneval, ak)
+		}
+	}
+	return unknown, uneval
+}
+
+// Check evaluates all critical and warning alert rules. An error is returned if
+// the check could not be performed.
+func (s *Schedule) Check(T miniprofiler.Timer, now time.Time) (time.Duration, error) {
+	select {
+	case s.checkRunning <- true:
+		// Good, we've got the lock.
+	default:
+		return 0, fmt.Errorf("check already running")
+	}
+	r := s.NewRunHistory(now, cache.New(0))
+	start := time.Now()
+	for _, ak := range s.findUnknownAlerts(now) {
+		r.Events[ak] = &Event{Status: StUnknown}
+	}
+	for _, a := range s.Conf.Alerts {
+		s.CheckAlert(T, r, a)
```

Comment:
I don't understand how unknowns work with recursive dependencies. Say we have C depends on B depends on A. A goes unknown. C is checked. It should be marked uneval, but I can't understand how it would be.